pkg/operation/botanist/etcd.go

```diff
@@ -58,14 +58,15 @@ func (b *Botanist) DefaultEtcd(role string, class etcd.Class) (etcd.Interface, e
 		b.Shoot.SeedNamespace,
 		b.SecretsManager,
 		etcd.Values{
-			Role:                    role,
-			Class:                   class,
-			Replicas:                replicas,
-			StorageCapacity:         b.Seed.GetValidVolumeSize("10Gi"),
-			DefragmentationSchedule: &defragmentationSchedule,
-			CARotationPhase:         v1beta1helper.GetShootCARotationPhase(b.Shoot.GetInfo().Status.Credentials),
-			KubernetesVersion:       b.Shoot.KubernetesVersion,
-			PriorityClassName:       v1beta1constants.PriorityClassNameShootControlPlane500,
+			Role:                        role,
+			Class:                       class,
+			Replicas:                    replicas,
+			StorageCapacity:             b.Seed.GetValidVolumeSize("10Gi"),
+			DefragmentationSchedule:     &defragmentationSchedule,
+			CARotationPhase:             v1beta1helper.GetShootCARotationPhase(b.Shoot.GetInfo().Status.Credentials),
+			KubernetesVersion:           b.Shoot.KubernetesVersion,
+			PriorityClassName:           v1beta1constants.PriorityClassNameShootControlPlane500,
+			TopologyAwareRoutingEnabled: v1beta1helper.IsTopologyAwareRoutingEnabled(b.Seed.GetInfo(), b.Shoot.GetInfo()),
```

Comment:
Since this might value be required multiple times, you could add it to https://github.com/gardener/gardener/blob/16a28f9ae567b5233b786dd72f28466414744183/pkg/operation/shoot/types.go#L66-L97 and only compute it once in the beginning of the flows.

Reply:
Yes, good point. I also didn't like that the test is duplicated for each and every component. In https://github.com/gardener/gardener/blob/16a28f9ae567b5233b786dd72f28466414744183/pkg/operation/shoot/types.go#L66-L97 there is no Seed object. The Seed object is needed to compute `TopologyAwareRoutingEnabled`. I will check how to add the Seed object.

Reply:
Yes. This was also how I started before you comment. I added a `*gardencorev1beta1.Seed` to the `shoot.Builder` and added a `TopologyAwareRoutingEnabled` field to the `shoot.Shoot` struct. In the `Build` function we set the `TopologyAwareRoutingEnabled` field. Maybe the only difference with your proposal is that I didn't add a `seedFunc` (similar to `cloudProfileFunc`) field to the `shoot.Builder` and directly added `*gardencorev1beta1.Seed` as field. See https://github.com/gardener/gardener/pull/7191/commits/861d5fa13f4cc77d61988e4ffff5f6a145297df6.

Reply:
OK, thanks, this works as well. Do you think we can follow the same for the `CloudProfile` to have a "streamlined approach" here, or is this not possible without additional effort?

Reply:
As far I see I cannot follow easily the same approach for the CloudProfile. In 1 place the CloudProfile is not already fetched and `WithCloudProfileObjectFromCluster` is used. To allow this usage, I think `cloudProfileFunc` is used in the builder instead a `cloudProfile` reference. For the Seed object, it is already fetched in the places where the builder is build, hence I used a `seed` reference instead of `seedFunc` in the builder.